### ISSUE TITLE
Proposal for strongly typed config and plugin system

### ISSUE THREAD

**Issue by alvrs** [latticexyz/mud#746]
This proposal is based on explorations for an extendable plugin system by @dk1a, see #679

## Terminology

- “Strongly typed”: In the context of Typescript we call a type “strong”, if it includes specific instances of primitive values instead of the general type. Example: a type `{ field1: "value", field2: 1 }` is strongly typed, `{ field1: string, field2: number }` is not.

## Motivation

- With a strongly typed config, we’re able to dynamically define types for `recs`, the upcoming composite datastore (#641), and more, without the need for code generation. A dynamic type definition is usually easier to maintain than logic for code generation, and errors can be caught more easily.

## Current Situation

- We rely on `zod` for config validation, which unfortunately does not support strong types (only weak primitive types like `string`, `number` etc are returned)
- Another requirement next to strong types is an extendable plugin system for the config, which got started in #679.
    - It is important that plugins are able to define their own config types, and expand/transform the existing config. This makes it much more complex to achieve strong config types.
    - Currently we solve the requirement of plugins extending the config types and transforming the existing config with an approach similar to Hardhat's, which involves overriding global type declarations and a global context, but unfortunately doesn’t support strong types (yet).

## Proposal

TLDR: [explore the interactive TypeScript playground for the explanations below](https://www.typescriptlang.org/play?#code/PQKhCgAIUg5B5AKgUQFyUQCwJYGdJ6QCGkADgE4D2pluRANpJQHaQAC9RALl9gMYBTAB4BPAF7AAtgFcAJgBpIuAQMgANDZABmlcpD4suwrsWazIsvH2m5c2FlBiOMOfBWq0GkAOaUB+SgA3AT0iVgYjcmZubGDiUnciPkxtXSVKekDsZm9IACNKLhTyAQBHaWwSyQFmLnwAClwuKhzILhFSf0UDZi1sXOMa2SI8+lU8kTJ6aW9s3ABKADpneHDIHG8UseDGItUevtzZAT7mbF4WfTD88aoAaxqLSgB3Vmy2zBLVJqJvf1RnM4AIyLFyqGwhCwnbL+D77FiHfJEZTmS7XAAGAElmKRpFx0W0OgJFM8cGNIA8BKRsrkmi1cu1OrhnAAmUFYeG9foEfDCUhhY7mLiUUyQdHIIT8swCWQExnEyCk7DkynU1p0lgMonM6CQADM7Mwqj5Apl+gR3MIgQY2GGRnMWRI6LElFlijyeLhSmamsJnShp3O9lYWiokm9v1ULJ5FKpXGWuucyGC5EmpGms1Yx1OsIx2Nx+MUGIlUsFAB5EJAALyQfN4gB8cqJ8mcAuIYpNZgAwhbvAStNJmHwLqxhZBJERSNow2K6-i2iLxZLTbKEzAACIiyxNGkVXApdMzObTyjhgRJFKFI16d6IIkAZT45GwpC4ij2kwYuC30OYqhIg7YOUqjoraBJ0jSa7AOA4DZJEWhJKoACycgAAoZtkZZztWpgiIoJYrjhYQiPWkAAN5QAQsjoBBOQANyUZ2sg9ly3joGWXaQIMZj4HO9b1Ac-ToF28zVqRBHSrIDEAL4waAEC6hy5olOarFMMw9CTHyAjDvgJATlOlBaFMR7MLga5gqpiKEJwkRccukn5JMfAMPQNJekxLGItURSukwxnnskziHpmBCsLoxx6EZ8SdEQ5BhIIYVegZoIAOrgso7YGWFTTnqixnXHF8V4c4eQ6UQEIECYE4PPg5xcci2CQmO8r6EafB3Kkeh7PZeBGEOqjRdpw4ysFGFmUlPUpZA9TwHkABWOkmA8IgBOQkUxt40hxWERhmnYA2QMg94sgADECACsijKAIziYDwpC4KgwDAM8b2LLlCHMPNch4IsBiSMAwrDCIAC02Cg2McV-rIwDuJ05DtKDEUhODuCgxQMr8FwIxjODzCg-NRDWrgT4vlwyMLUt6MHYIoP+Gdl3QdA0HgK16GmfgNYAEo6RFZa0d4igobIHOZvWDGtV53I1mRJmZk9kBi8e0kMeA8nOIgIrKtMdLcKoPU2JGijXqohDXMIRCSOmqg6KESvjZAACCLa6kqyRtEQtXkdoTX0LIQI0c07nScQ+CCbk2QFq7MBtkxdUmBM5qkCI7k9X0Aj+0CkDWtMBsikxMoAGJ+wHAIszBrVzk7OFy+AACQGf+07QfPjkLb1yFczoBR9f1y9kAAEKTDV7nK8wNf8oj-keZKS1mlHnryh3-fAKpuDSNU5D4CUgixP+frGuQVA3sZPVd28+DMIUYV8NMgoN6vk1GlZ-SP+PLcO6ZTsN7JquV0SI6DlBROwrDhaupEayVm4rIXiOI8Q1wbgAflrPArgNcABkPtC6yBLpnWQn9EAAG0ABETcCEkIALqQFDg3dAf4Uxqx6E0L+mYa6y0oradAJDvzVFBhvBGoNALAQxuNcGsgSGu3sqWaWbFtCDmHMGSAYCYFwILE7fii8uDoEQKJCikADGQBKFwaQURyKPwMYsKxWjFAWOkSuPBzd0BaMWOQn+9cDGh2RJAQcdxr6vDDkA0sMpQGIHrJRP+gTmEmDoLwXAGd8Ai3HlhNBLsgkrlARAiWAD-QfxwvKaKH81Ya0UiKSq49nZFnXpvSEBhpD+wDDCOElRX65G8Toeg9AXhPUcKzaJrTs4cMbqXT+PDTwCFBrnaQAhJENwvorXu9cP6-0ogPWAhQQLeOieiEkOAPYmKiOHXsYUxwlCILIFgmlVlrzbOc2B3p6SH2ZF4o5ZkYkxHiU1NRCDilgGcE7egkRoi8B2JMZ48J6nmHcFkY47YjT0ARvIocI4FzEDxKeGILlOmTGyFoEIzgerDXtGFAsh9roihIMwmpehniUEhcwQAmAQmEwMTfOBBra6BMNXSAoNFRklUKqdyGpWjyh6RXdWa8AASmdEWtRaqY1g6Ia4AB8h4EjgouTBaq2aAIAKpnBYFrbEkRlCKJYGWXVkCZq6vsv1e5xFIAoPqHcdAurRJVlIoESgtpID0IECmUSqiZourCnivQmJ3Weu9eYFBmJfWQAYSEYpUqZXNUAWOXFzUX5aKeSecMJB5mtMCYBS49RuWqrnIPeYOrcnjVwHOMsqFbVDHwOPXAVrUI7R8iEXAjaiErUKZQ0hnlexUPrEQ06lDk2oLDV6C+JKl7akCXBHtS0lFltSZALBlbq2tSQiEP4os60NqbUGttVr9XBiNbUVdZrmCNuPWgxt9ZskSsgNKhFkIBzIqUWObMTSeoRz5cqAVKg1S0h9CK7UywB7pSYJIeqPV0TzIJCtPNXoSjlEqGaIDZUXKVXqngN9-6YZKCpDtIwmln6EHhYi79d7lj0ZRSRgQsjG3Np4qwuYihOJBvgAhrgZZ93kEPW2k99ZFAkPmSQl99Q5l1vQKhDuEdhLgHmMJLdPt5mKZoeYp+cHsh3zkAbF+hb3g9TpVwUl8oG7GIVT7Kx-1eyKELX-OSfzdTrOeNUre4drgseaXoIDbkHgfDwOXEArNYOqBZXEVqdA8XtGfpyREJGgwsBbAPNsPUSgb0Be5IDlAqbDnWN44VWomTgH6RHaMNYWOyPqHLPJ0lFByzcdw3hAgABqDBpkkN01slgTR5i-IUjAJ2thN7uXBcQWQ5gwhXkhBUweuz+ApCJfgchg96GbzKqEOqHLEaOd6Tk1Qlba6+3wdthNu3IT-yq0Nkw49B6104dRSAJD-U1FBpIXQEzhHTNEaZcRsyDEjtYugJjSiVFCDtd8rgg9NFoJ0XoyiBi7NmK0QxTx4AXnr3eXEhJkAknjRSQWFbqDyevpKTARzPU0sotpZCyAXTKCdRCyZ8LJ231wdiwbQBCWBBJfMy-IDDPgyZZuWYDD-h6m8FaIV4rzKyuQYq-4B7bzWl6hwvV3scn651z7h-Fez3f4d0We1j7nXJm9ZmYoAeiBUxtTCH8ML+AxydaKO5Rb3UWVrEFqYcwzxuAe0QPeJQT5zz5pMCIOlj8tvoGzr-KJj21OjecF5os+Bqi1GDGac85A3IhEUCwM2h2ca1CMWUCoVQagmEID1EYypziTDtvLOYBKC7AK9EB+oSc7nuV+ypchuBFBMSFZgTl1gTAj9RXsZwA5OmNLOMKMy10hfy6OCcCqgKc629H1xLgfB5hFml23mwqKJxwSIO8crTyY5cWCKwZ45wUi+UynFPI5x4pO87HYIb5KTAew3UgC4uQ2VwWYVIQw6kXoi+VGOCrSFkzg0qKkNGLwio+whUyQTUcQ5wEWUWKan6egUOlwY4k46Yn4S+6I4Ohw-YCiI4zIDuD4ZMr4EBAA5PXr0JCCQHfq1KGKeG7u6J6DNi5MwG+juEvheDgWbInJMGlq0CQLRmmv6GVF0s8OACQawDQf0BxBxvcrIvxCppACJOYgYmMCYAgTWBHNjl1DNP0vOtFHNItMOIsFMv4AJL2IsPMvMKjoYvYo5LLJAI5jgooI5hfIsNod4B4axKJP-DjujkLvZjgjJHJIQbKumiKAcv7qrlRvKOYEBhqu2CWuFFoG+q5B2MArInQT+uASxuYCLilm-FLKeF-n+BJMxL2BxFarxrDi2lpgpqGktnWjQpRCgm2nofgLzAYOtALMHO3MTmhONOEn4WMXWv2gILHsZG2kOp9pUaOpQmjvGomuQJRH6owqke+qmiAf6GOMYPFCVm2FvK7jlokWYqKjPCQLME-j4gaiUUinesyHugegILzFkXeJ0DDnDjNI5nFN4IrMRJGrhBer8deiamuuaqCQquCQIBWC+tOqDASUAlbDbJALqnQK7gSaDBcUBiyvgLwYAm1iMtwrnJIpdv7NdkCC5gMU1uNGks9rprJNVscnVr+Kxnrjyd-IoPyS1j7JbmMnwm4ayQnpANnHjtEiNhcV2GckYO2CmP-n8dQXsaxASC-kUA8r6KKpLuhu7CkCIdcJmr7qoE8ftKrlXlkbmm+tFOUYaTIr2NUQCRKm+nke0bIqVu7kSACAPLidEYcCjmJOkpJKEqRKqtGUYbovGe0TKIPLiRriwsGUaYiDWJEYEj2JIK0QICGZ0QUsZBHNkgPJmR0WpLSYfOgG+g2UmZpg2dmWEm+oEXKdboqWPsAsXEyVbuMjbnnEqaXBySMa1A2aGTWMJn8JiVENiWWNWYfLIJWaxOuUSNFLWdkkAA)

This proposal goes over an alternative approach for solving both requirements (strong types, config extendable by plugins). On a high level the config definition can be broken down in three stages:

1. The user defines the config based on an `Input` type, while keeping strong types
2. The config is expanded to an `Expanded` type, while keeping strong types
3. The expanded config is validated via `zod`, but the strong type definition from stage 2 is kept.

Every plugin defines an `Input`, an `Expanded<T = Input>` type, and a `expandConfig` function to map from `Input` to `Expanded`. Do distinguish plugins from each other in TypeScript, they also define a unique `id` string.

```tsx
interface MudPlugin<Input = any, Expanded = any> {
  id: string;
  expandConfig: <C extends Input>(config: C) => Expanded;
}
```

The core config only expects a map of plugins. The config is later expanded by calling the `expandConfig` method of each plugin in order of appearance in the map. We use a map instead of an array, because it makes it easier to type check for the existence of expected plugins in the map. ([[Object keys order is guaranteed since ES2015.](https://www.stefanjudis.com/today-i-learned/property-order-is-predictable-in-javascript-objects-since-es2015/)](https://www.stefanjudis.com/today-i-learned/property-order-is-predictable-in-javascript-objects-since-es2015/))

```tsx
type Plugins = Record<string, MudPlugin>;
type Config = { plugins: Plugins };
```

To illustrate the usage, here is an example for a Plugin A, which takes `{ field1: string }` as config input, and expands it by copying the `field1` value to `expandedField1`:

```tsx
type InputA = { 
  fieldA: string,
  plugins: {
    // By making PluginA part of the expected input type,
    // consumers receive a type error if the plugin is not included
    // in the config
    PluginA: PluginA
  }
};

type ExpandedA<T = InputA> = T extends InputA 
  ? InputA & { expandedFieldA: T["fieldA"] } 
  : never;

const PluginA = {
  id: "some-super-unique-plugin-id",
  expandConfig: function <T extends InputA>(input: T) {
    return {
      ...input,
      expandedFieldA: input.fieldA
    } as unknown as ExpandedA<T>
  }
} as const satisfies MudPlugin<InputA, ExpandedA<InputA>>;

type PluginA = typeof PluginA;
```

To use Plugin A, a consumer could define their config as follows:

```tsx
import { PluginA, InputA } from "plugin-a";

const pluginConfig = {
  fieldA: "some-value",
  plugins: {
    PluginA
  }
  // Note `as const`, which turns config into readonly
  // and adds strong types
} as const satisfies InputA;
```

Alternatively we could provide a helper function to automatically infer the expected input type, so a consumer wouldn’t have to import `InputA` - while keeping strong types:

```tsx
// Helper type to turn `A | B` into `A & B`
type UnionToIntersection<U> = (U extends any ? (k: U) => void : never) extends (k: infer I) => void ? I : never;

// Helper type to infer the input types from a plugins config as union (InputA | InputB)
type PluginsInput<P extends Plugins> = Parameters<P[keyof P]["expandConfig"]>[0];

// Infer the plugin input types as intersection (InputA & InputB)
type MergedPluginsInput<P extends Plugins> = UnionToIntersection<PluginsInput<P>>;

// Helper function to define the config while keeping strong types.
// We omit the `plugins` key from the required config because it is
// defined separately in this helper function.
function defineConfig<P extends Plugins, C extends Omit<MergedPluginsInput<P>, "plugins">>(
  plugins: P,
  config: C
): C & { plugins: P } {
  // We include the plugins in the output type
  return { ...config, plugins }
}
```

Now consumers can define their config like this:

```tsx
import { PluginA } from "plugin-a";

// We have type safety in the config definition,
// and the resulting config object has strong types
const config = defineConfig({ PluginA }, { fieldA: "someValue" } as const);
```

Assuming we add another Plugin B, which expects `fieldB: number` as import, the definition would look like this:

```tsx
import { PluginA } from "plugin-a";
import { PluginB } from "plugin-b";

// We have type safety in the config definition,
// and the resulting config object has strong types
const config = defineConfig(
  { 
    PluginA,
    PluginB
  },
  {
    fieldA: "some-value",
    fieldB: 1
  } as const
);
```

Now, as mentioned earlier, one important requirement is the ability for plugins to expand the config (by adding more fields, expanding shortcut fields to the full definitons, setting default values, etc), and for us to maintain strong types, even with those arbitrary expansions, so other type definitions can depend on the fully expanded config.

Here is how we can achieve it:

```tsx
// Helper function to apply all `expandConfig` functions
// TypeScript can't infer a strong type from this, but we can
// still achieve it by defining a helper type below
function expandConfig<C extends Config>(config: C) {
  let expanded = config;
  for (const plugin of Object.values(config.plugins)) {
    expanded = { ...expanded, ...plugin.expandConfig(config) };
  }
  return expanded;
}

// Helper type to turn a strongly typed config into a union of
// all `expandConfig` functions defined in the config
type CombineExpandConfig<C> = C extends { plugins: infer Plugins }
  ? Plugins extends Record<string, MudPlugin>
    ? Plugins[keyof Plugins]["expandConfig"]
    : never
  : never;

// Helper type to extract and merge the return types of a given union of functions
type MergeReturnType<T extends (...args: any) => any> = UnionToIntersection<ReturnType<T>>;

// --- Example Usage ---

// config has strong type { fieldA: "val", fieldB: 1, plugins: { PluginA, PluginB } }
const config = defineConfig({ PluginA, PluginB }, { fieldA: "some-value", fieldB: 1 } as const);

// Create a version of `expandConfig` with strong types,
// from which we can infer the merged strong return types 
// of all `expandConfig` functions
//
// typedExpandConfig has type:
// <T>(config: T) => ExpandedA<T> | <T>(config: T) => ExpandedB<T>
const typedExpandConfig = expandConfig as CombineExpandConfig<typeof config>;
// ExpandedConfig has type: 
// ExpandedA<T> & ExpandedB<T>
// = { fieldA: "some-value", expandedFieldA: "some-value", fieldB: 1 }
type ExpandedConfig = MergeReturnType<typeof typedExpandConfig<typeof config>>;

```

### Additional considerations

Our current Store config definitions uses generic types to add additional type checks (to allow user defined Enums to be used as schema types and primary key types). To keep those additional type checks, the Store plugin can export its own `defineStoreConfig`, which just passes the config through, but adds type checks as before:

```tsx
// Do arbitrary type checks, but return the strongly typed input 
function defineStoreConfig<T>(config: T): T {
  return config;
}

const config = defineConfig(
  { 
    StorePlugin,
    WorldPlugin
  },
  { 
    // the Store plugin's config definition helper function can be
    // used inside the main helper function for additional type hints
    ...defineStoreConfig({ ... }),
		
    // For plugins that don't require their own additional type hints
    // we can just put the config options directly
    worldContract: "CustomWorldContract",
    // ...
  });
```

**Comment by dk1a** [comment]
I want to separate 2 concerns here:
1. expanding the user config while keeping strong types
2. unglobaling plugins, making them an option of the config itself

the 1st part is compatible with the current setup, I basically forgot about it and add soon. The 2nd is more invloved and would much more significantly change the current approach to the config

**Comment by holic** [comment]
Just came across this article:
https://devanshj.me/writings/the-four-declaration-pattern-for-typescript

Not sure if it's useful here but it reminded me of our plugin system and combining types

**Comment by dk1a** [comment]
- We have fully global plugins currently
- https://github.com/latticexyz/mud/pull/1077 explores fully local plugins
- Local types + global data is the worst combination, since local types are the hardest part and we should go fully local if we can do them well
- Global types + local data remains unexplored, I provide some thoughts here

(I list cons, the pros are just an absence of another's cons; makes it easier to compare I think)

### Fully global

1. Need 2 separate package exports for "library without side-effects" and "extender with side-effects"
2. Import graph is easy to mess up (the current implementation likely has minor bugs, or weird behavior at least)
3. Interface can't be inferred from a config extender, it needs a manual declaration merge
4. [Declaration merging](https://www.typescriptlang.org/docs/handbook/declaration-merging.html) is an obscure feature, coming in people probably won't know it

```typescript
import { mudConfig } from "@latticexyz/world/register";
import "@latticexyz/world/snapsync";

export default mudConfig({
  snapSync: true,
  tables: {
    CounterTable: "uint32"
  }
});
```

### Fully local

1. The ideal version requires a feature typescript doesn't have (higher-kinded types)
2. The implementable version relies on hacks (type expansion hidden in userland, verbose boilerplate in tests)
3. Can have slightly worse type errors than interface merging
4. Can be less powerful, since the type/interface is never defined, only inferred (I don't have examples and not sure this is relevant)

```typescript
import { mudConfig } from "@latticexyz/world";
import { snapsyncPlugin } from "@latticexyz/world/snapsync";

export default mudConfig({
  plugins: { snapsyncPlugin },
  snapSync: true,
  tables: {
    CounterTable: "uint32"
  }
});
```

### Global types + local data

1. All the downsides of `Fully global` (type side-effects remain, which are a worse superset of data side-effects) 
2. Types and data diverge if the plugin is imported but not used
3. The import now has `from`, which may or may not make imports even easier to mess up ([example of confusing stuff](https://github.com/latticexyz/mud/blob/main/packages/store/ts/register/index.ts) where using `export { mudCoreConfig, resolveTableId } from "@latticexyz/config/register";` instead of 2 lines would break things)

```typescript
import { mudConfig } from "@latticexyz/world";
import { snapsyncPlugin } from "@latticexyz/world/snapsync";

export default mudConfig({
  plugins: { snapsyncPlugin },
  snapSync: true,
  tables: {
    CounterTable: "uint32"
  }
});
```

**Comment by holic** [comment]
done!
